### ISSUE TITLE
Various bugs with UpdateDateColumn and CreateDateColumn and Postgres

### ISSUE THREAD

**Issue by mothershipper** [typeorm/typeorm#7150]
## Issue Description

Both the UpdateDateColumn and CreateDateColumn decorators are broken in some way for Postgres.

CreateDateColumn has no value being set on save, UpdateDateColumn does not work if the column has a transformer set on it.

### Expected Behavior

On save, I'd expect TypeORM to set a createdAt date.

When I have a transformer, I'd expect TypeORM to run a new date through the transformer -- or just work?

### Actual Behavior

No dates are set automatically for non-nullable columns.

When I have a transformer, TypeORM makes no attempt to update the updatedAt date.

### Steps to Reproduce

Both bugs are outlined here:

https://github.com/mothership/typeorm-bug

### My Environment

| Dependency          | Version  |
| ---                 | ---      |
| Operating System    |    OSX      |
| Node.js version     | v12.18.3 
| Typescript version  | v4.2.1 | 
| TypeORM version     | v0.2.29 | 

### Relevant Database Driver(s)

- [ ] `aurora-data-api`
- [ ] `aurora-data-api-pg`
- [ ] `better-sqlite3`
- [ ] `cockroachdb`
- [ ] `cordova`
- [ ] `expo`
- [ ] `mongodb`
- [ ] `mysql`
- [ ] `nativescript`
- [ ] `oracle`
- [x] `postgres`
- [ ] `react-native`
- [ ] `sap`
- [ ] `sqlite`
- [ ] `sqlite-abstract`
- [ ] `sqljs`
- [ ] `sqlserver`

### Are you willing to resolve this issue by submitting a Pull Request?

- [ ] Yes, I have the time, and I know how to start.
- [ ] Yes, I have the time, but I don't know how to start. I would need guidance.
- [ ] No, I don't have the time, although I believe I could do it if I had the time...
- [x] No, I don't have the time and I wouldn't even know how to start.

**Comment by nebkat** [comment]
Are you synchronizing the tables in the DB? `CreateDateColumn` relies entirely on the default value being set in database.

**Comment by mothershipper** [comment]
@nebkat that's a good callout -- the migrations were done by hand, so the `CreateDateColumn` did not have a default set in the db. 

That solves one of issues, however, we're still having trouble getting the `UpdateDateColumn` to work when we have a transformer set.  It appears to work properly without one.

**Comment by nebkat** [comment]
Is the `UpdateDateColumn` value not being stored in the database, or is it just not being returned?

If you provide a value for the update date column it isn't automatically set to `CURRENT_TIMESTAMP`, it's just treated as a normal column, so you should follow normal debug procedures and check the SQL output.

**Comment by mothershipper** [comment]
So, I updated the example -- we now set a default value of `now()` in our migration for both the `createdAt` and `updatedAt` columns:

```
CREATE TABLE posts (
    id integer PRIMARY KEY,
    content TEXT NOT NULL,
    created_at TIMESTAMPTZ NOT NULL DEFAULT now(),
    updated_at TIMESTAMPTZ NOT NULL DEFAULT now()
);
``` 
Our code no longer sets either `createdAt` or `updatedAt`

```typescript
  
const unixTimestamp: ValueTransformer = {
  from: (value: Date | null | undefined): number | null | undefined => {
    return value ? value.getTime() : value
  },
  to: (value: number | null | undefined): Date | null | undefined => {
    return typeof value === 'number' ? new Date(value) : value
  },
}


@Entity({ name: "posts" })
class Post {
    @PrimaryColumn({
        nullable: false,
        primary: true,
    })
    id!: number;

    @CreateDateColumn({
        name: "created_at",
        type: "timestamptz",
        transformer: unixTimestamp,
    })
    createdAt!: number;

    @UpdateDateColumn({
        name: "updated_at",
        type: "timestamptz",
        transformer: unixTimestamp,
    })

    updatedAt!: number;

    @Column({
      name: 'content',
      type: 'text'
    })
    content!: string
}


const delay = (ms: number): Promise<void> => {
  return new Promise((res) => {
    setTimeout(res, ms)
  })
}


const main = async () => {
  const connection = await createConnection({
    type: "postgres",
    url: "postgres://test:test@localhost:5432/test",
    entities: [ Post, PostNormal ],
  });

 const postRepo = connection.getRepository(Post)
  const post = postRepo.create({
    id: 3,
    content: 'Hello!',
  })
  console.log("Saving createdAt:  ", post.createdAt)
  console.log("Saving updatedAt:  ", post.updatedAt)
  const savedPost = await postRepo.save(post)
  console.log("Saved createdAt:   ", savedPost.createdAt)
  console.log("Saved updatedAt:   ", savedPost.updatedAt)
  savedPost.content = "world"
  await delay(1000)
  const updatedPost = await postRepo.save(savedPost)
  console.log("Updated createdAt: ", updatedPost.createdAt)
  console.log("Updated updatedAt: ", updatedPost.updatedAt)
}

main()
```

We get the following logs:

```
Saving createdAt:   undefined
Saving updatedAt:   undefined
Saved createdAt:    1607100629633
Saved updatedAt:    1607100629633
Updated createdAt:  1607100629633
Updated updatedAt:  1607100629633
```

And see the following in the db:

```
test=# SELECT * FROM posts WHERE id = 3;
 id | content |         created_at         |         updated_at
----+---------+----------------------------+----------------------------
  3 | world   | 2020-12-04 16:50:29.633+00 | 2020-12-04 16:50:29.633+00
(1 row)
```

So, TypeORM does appear to be reading back the `updatedAt` column, and it is saving the updated record (content changed to 'world'), but it does not automatically update the `updatedAt` column.

**Comment by jbjhjm** [comment]
@mothershipper 
I'm searching for a solution too, using mariaDB. I think this is not database-specific but a general/conceptual issue.

My guess is the transformer will only be called if you update an Entity with manually specified updateDate.
If you rely on trusting TypeOrm to do this automatically in the background, the to-transformer will not be called.
On the other hand, because the column will be loaded the same way as all other columns, the from-transformer will be used.

I did not check all code/execution order, but here's where the updateDate will be added when using queryBuilder:
https://github.com/typeorm/typeorm/blob/master/src/query-builder/UpdateQueryBuilder.ts#L478

**Comment by mothershipper** [comment]
@jbjhjm / @nebkat / @pleerock  I did a bit more digging into this:

In the first test case -- using Date / no transformer, the query logs show `updatedAt = CURRENT_TIMESTAMP`, which is the codepath @j6lim identified.

```sql
START TRANSACTION
UPDATE "posts" SET 
  "content" = $2,
  "updated_at" = CURRENT_TIMESTAMP 
WHERE 
  "id" IN ($1) 
RETURNING "updated_at" -- PARAMETERS: [2,"world"]
COMMIT
```

However, in the second test case, we see that typeorm IS updating the timestamp, but it's updating it with the previous value -- this suggests that TypeORM has detected a change to the updatedAt column (because we transformed it), and is saving the previous value loaded from the database. 

```sql
START TRANSACTION
UPDATE "posts" SET 
  "created_at" = $2, 
  "updated_at" = $3, 
  "content" = $4 
WHERE 
  "id" IN ($1) 
RETURNING "updated_at" -- PARAMETERS: [3,"2021-04-21T19:22:29.785Z","2021-04-21T19:22:29.785Z","world"]
COMMIT
```

Note the relevant code:

```js
if (metadata.updateDateColumn && updatedColumns.indexOf(metadata.updateDateColumn) === -1)
   updateColumnAndValues.push(this.escape(metadata.updateDateColumn.databaseName) + " = CURRENT_TIMESTAMP");
```

TypeORM only adds the `updatedAt = CURRENT_TIMESTAMP` IF the column is an updateDateColumn and it has NOT already added it to the sql expression.  Since the column is in the SQL expression, that points at the transformer confusing TypeORM when it comes to detecting changes on save.

**Comment by mothershipper** [comment]
Ok, a bit further now --

I think this is the essence of the problem:

```js
const time = 1619035148
const date1 = new Date(time)
const date2 = new Date(time)

console.log(date1 === date2) // logs false
```

TypeORM handles this actually, by normalizing date columns and entity values specially.  It turns the date column into strings for comparison:

https://github.com/typeorm/typeorm/blob/7eb032705912cbf4ee340ed9e49970d0f6e23714/src/persistence/SubjectChangedColumnsComputer.ts#L77-L123

But... if the column has a transformer, it applies the transformer AFTER it tried to handle the special-case comparisons:

https://github.com/typeorm/typeorm/blob/7eb032705912cbf4ee340ed9e49970d0f6e23714/src/persistence/SubjectChangedColumnsComputer.ts#L125-L127

So that when it finally compares the new / old values:

https://github.com/typeorm/typeorm/blob/7eb032705912cbf4ee340ed9e49970d0f6e23714/src/persistence/SubjectChangedColumnsComputer.ts#L130-L132

It detects a difference.

TL;DR; Without the transformer, this code is comparing timestamp strings for equality.  With the transformer it seems that it's comparing a date object with a timestamp string.

@pleerock / @nebkat would it be reasonable to transform the value prior to doing special-case handling?  I can PR that fix :)

**Comment by mothershipper** [comment]
I believe this #7588 resolves this issue -- but thinking on it further, there may be a need for a better fix for transformers in general.  

Because TypeORM is doing a `===` to see if a column has changed, any transformer that returns a class or non-primitive data type is going to be detecting changes to columns that probably haven't changed.  

Not sure if the transformer is the appropriate place to define the equality check, but say we performed the equality check by transforming the db value to the expected entity value type, then define an equality comparison that way:

```typescript
const unixTimestamp: ValueTransformer = {
  from: (value: Date | null | undefined): number | null | undefined => {
    return value ? value.getTime() : value
  },
  to: (value: number | null | undefined): Date | null | undefined => {
    return typeof value === 'number' ? new Date(value) : value
  },
  equal: (left: number | null | undefined, right: number | null | undefined) => {
    return left === right
  }
}

const jsonTransformer: ValueTransformer = {
  from: (value: string | null | undefined): object | null | undefined => {
    return value ? JSON.parse(value) : value
  },
  to: (value: object | null | undefined): Date | null | undefined => {
    return value ? JSON.stringify(value) : value
  },
  equal: (left: object | null | undefined, right: object | null | undefined) => {
    return deepCompare(left, right)
  }
}
```

This would definitely avoid unnecessary updates, but is a potentially breaking API change. But the approach is more flexible than TypeORM trying to normalize every potential return-type of transformers.

**Comment by mothershipper** [comment]
@imnotjames saw you added the PG label, but I think this affects more than just postgres -- was able to fix in the linked PR, but the bug wasn't related to anything postgres specific

**Comment by imnotjames** [comment]
> @imnotjames saw you added the PG label, but I think this affects more than just postgres -- was able to fix in the linked PR, but the bug wasn't related to anything postgres specific

Good to know!  The `postgres` label gets added via a script because of the entries checked above.

**Comment by mothershipper** [comment]
@imnotjames - really appreciate the work you've been doing here, it's great to see some velocity on the project :). 

No rush on this, but wondering if you had a chance to look at the [associated PR](https://github.com/typeorm/typeorm/pull/7588)?  The PR goes the less-breaking-change route, and just attempts to apply the column transformer when comparing values. It's been a moment since I've looked at the code, but from what I recall, any column using a transformer would see `UPDATE` queries on `save()`, even when no value has changed -- this issue just happened to find this behavior on an `UpdateDateColumn`.

I still think the approach (or something similar) outlined in [this comment](https://github.com/typeorm/typeorm/issues/7150#issuecomment-824460389) is worth considering, but would be a reason for a major version bump (I guess minor version possibly, since TypeORM is not at 1.x.x?) .

**Comment by pleerock** [comment]
Your PR shows design issue we have with `CreateDateColumn` / `UpdateDateColumn`. Steps to reproduce issue:

1. on initial `save` you rely on initial `UpdateDateColumn` value inserted by database (`CURRENT_TIMESTAMP`).
Then your entity value is populated by date returned by database.
2. you change any other entity property
3. you `save` entity again

What happens next? `updateDate` definitely must be updated in the database, because change in other property has occur. But how is it updated? Using `CURRENT_TIMESTAMP`? Looks like no, because we do allow **custom values** for columns decorated by `@UpdateDateColumn`. And after we save entity for the first time, we have a custom value inside that entity. But if custom value is the same as old value - what does it mean? Do we fallback to `CURRENT_TIMESTAMP` (because we MUST update column)? Yes, it falls back to `CURRENT_TIMESTAMP`, but only because of the comparison there of old column value in the db vs new column value (but in your example it doesn't because transformer breaks everything).

Very very confusing behavior and obviously bad design.
And now, using transformers things being complicated even more.

Best way for people to prevent this confusion - remove confusing @CreateDateColumn and @UpdateDateColumn decorators and show people how to do same using @Column decorator.

**Comment by sammous** [comment]
This is still an active issue ⚠️ , we faced the same things with the following transformer : 
```
export const myTransformer: ValueTransformer = {
    to: (date) => date,
    from: (date: Date) => {
        return isNil(date) ? date : formatMyDate(date);
    }
};
```
When an object if fetched from the DB, updated and then saved, it will consider the previous date for the `UpdateColumn` since it's not empty instead of overwritting this date.

A quick fix is to be forced to have a different transformer for specifically `UpdateColumns`. For instance if you want to have the columns updated with the current timestamp (which I imagine is the default behaviour you looking for) :

```
export const myTransformer: ValueTransformer = {
    to: (date) => Date.now(),
    from: (date: Date) => {
        return isNil(date) ? date : formatMyDate(date);
    }
};
```

**Comment by pilaoda** [comment]
The real problem I think is the `databaseValue` do `transformTo` first before `mixedDateToUtcDatetimeString`, while the `entityValue` only do `transformTo` without `mixedDateToUtcDatetimeString`.
Therefore two simple ways to fix:
- Remove the `mixedDateToUtcDatetimeString` for `databaseValue` if column has transformer.
- Add `mixedDateToUtcDatetimeString` after transform for `entityValue`.

Choose one way and we get consistency for diff compute.

**Comment by danLeBrown** [comment]
@mothershipper, did you eventually find a fix for this? I'm experiencing a similar issue and out of my wits here. I'm thinking of switching to MySQL since I've never experienced the same with it, and there are hooks like the `onUpdate` that work specifically for it.
